### PR TITLE
Organise unit tests to use pytest fixtures (`test_repository.py` implementation)

### PR DESCRIPTION
**Pull request by SeanBryan51**

### CODE REVIEW

File: tests/test_repository.py [hunk 1]
```diff
@@ -3,315 +3,353 @@
...
-    )
+    def test_standard_output(self, repo, mock_subprocess_handler):
+        """Success case: test standard output."""
+        with contextlib.redirect_stdout(io.StringIO()) as buf:
+            repo.checkout()
+        assert (
+            buf.getvalue()
+            == f"Successfully checked out trunk at revision {mock_subprocess_handler.stdout}\n"
+        )
+        with contextlib.redirect_stdout(io.StringIO()) as buf:
```
Comment (ccarouge) [inline tests/test_repository.py:72]:
```suggestion
       # Verbose output
        with contextlib.redirect_stdout(io.StringIO()) as buf:
```
It's hard to see why the test is repeated otherwise.

Or parameterise on verbosity...

File: tests/test_repository.py [hunk 2]
```diff
@@ -3,315 +3,353 @@
...
-    with contextlib.redirect_stdout(io.StringIO()) as buf:
+
+    @pytest.fixture(autouse=True)
+    def _setup(self, repo):
+        """Setup precondition for `CableRepository.pre_build()`."""
+        (internal.SRC_DIR / repo.name / "offline").mkdir(parents=True)
+        (internal.SRC_DIR / repo.name / "offline" / "Makefile").touch()
+        (internal.SRC_DIR / repo.name / "offline" / "parallel_cable").touch()
+        (internal.SRC_DIR / repo.name / "offline" / "serial_cable").touch()
+        (internal.SRC_DIR / repo.name / "offline" / "foo.f90").touch()
```
Comment (ccarouge) [inline tests/test_repository.py:114]: Should we add a teardown?

File: tests/test_repository.py [hunk 3]
```diff
@@ -3,315 +3,353 @@
...
-    # Success case: test non-verbose standard output
-    (tmp_dir / internal.CABLE_EXE).touch()
-    repo = get_mock_repo()
-    with contextlib.redirect_stdout(io.StringIO()) as buf:
+
+    @pytest.fixture(autouse=True)
+    def _setup(self, repo):
+        """Setup precondition for `CableRepository.post_build()`."""
+        (internal.SRC_DIR / repo.name / "offline" / ".tmp").mkdir(parents=True)
+        (internal.SRC_DIR / repo.name / "offline" / ".tmp" / internal.CABLE_EXE).touch()
```
Comment (ccarouge) [inline tests/test_repository.py:236]: Teardown?

File: tests/test_repository.py [hunk 4]
```diff
@@ -3,315 +3,353 @@
...
+        }
+
+    @pytest.fixture(autouse=True)
+    def _setup(self, repo, netcdf_root):
+        """Setup precondition for `CableRepository.run_build()`."""
+        (internal.SRC_DIR / repo.name / "offline" / ".tmp").mkdir(parents=True)
+
+        # This is required so that we can use the NETCDF_ROOT environment variable
+        # when running `make`, and `serial_cable` and `parallel_cable` scripts:
+        os.environ["NETCDF_ROOT"] = netcdf_root
```
Comment (ccarouge) [inline tests/test_repository.py:181]: Teardown?

File: tests/test_repository.py [hunk 5]
```diff
@@ -3,315 +3,353 @@
...
+    @pytest.fixture()
+    def build_script(self, repo):
+        """Create a custom build script and return its path.
 
+        The return value is the path relative to root directory of the repository.
+        """
+        _build_script = internal.SRC_DIR / repo.name / "my-custom-build.sh"
+        _build_script.parent.mkdir(parents=True)
+        _build_script.touch()
+        return _build_script.relative_to(internal.SRC_DIR / repo.name)
```
Comment (ccarouge) [inline tests/test_repository.py:272]: Teardown?

File: tests/test_repository.py [hunk 6]
```diff
@@ -3,315 +3,353 @@
...
+        with pytest.raises(
+            FileNotFoundError,
+            match=f"The build script, {build_script_path}, could not be "
+            "found. Do you need to specify a different build script with the 'build_script' "
+            "option in config.yaml?",
+        ):
+            repo.custom_build(modules)
+
+
+class TestRemoveModuleLines:
```
Comment (ccarouge) [inline tests/test_repository.py:344]: Too bad this is preventing the `repo` fixture from being on auto-use. It's probably a sign this function should be moved to `utils`. But that's another issue.

File: tests/test_repository.py [hunk 7]
```diff
@@ -3,315 +3,353 @@
...
-        "mkdir src/trunk/offline/.tmp\n"
-        "cp -p src/trunk/offline/foo.f90 src/trunk/offline/.tmp\n"
-        "cp -p src/trunk/offline/Makefile src/trunk/offline/.tmp\n"
-        "cp -p src/trunk/offline/parallel_cable src/trunk/offline/.tmp\n"
-        "cp -p src/trunk/offline/serial_cable src/trunk/offline/.tmp\n"
+        tmp_dir = internal.SRC_DIR / repo.name / "offline" / ".tmp"
+        assert (tmp_dir / "Makefile").exists()
+        assert (tmp_dir / "parallel_cable").exists()
+        assert (tmp_dir / "serial_cable").exists()
+        assert (tmp_dir / "foo.f90").exists()
```
Comment (ccarouge) [inline tests/test_repository.py:123]: Why did you remove:
```python
shutil.rmtree(tmp_dir)
```
Are we not cleaning up anymore?

File: tests/test_repository.py [hunk 8]
```diff
@@ -3,315 +3,353 @@
...
+                "cp -p src/trunk/offline/serial_cable src/trunk/offline/.tmp\n",
+            ),
+        ],
     )
-    shutil.rmtree(offline_dir / ".tmp")
+    def test_standard_output(self, repo, verbosity, expected):
+        """Success case: test standard output."""
+        with contextlib.redirect_stdout(io.StringIO()) as buf:
+            repo.pre_build(verbose=verbosity)
+        assert buf.getvalue() == expected
```
Comment (ccarouge) [inline tests/test_repository.py:146]: No more clean up either...

**Review by ccarouge — CHANGES_REQUESTED** [review]
It's more a question than a request for changes. I see we aren't cleaning up created files and directories anymore. Is that a choice?

File: tests/test_repository.py (hunk 1 above)
Comment (SeanBryan51) [reply tests/test_repository.py:72]: Good spotting, looks like I forgot to parametrise this one...

File: tests/test_repository.py (hunk 2 above)
Comment (SeanBryan51) [reply tests/test_repository.py:114]: I figured all directories under the `mock_cwd` are removed at the end by the `_run_around_tests` fixture which is why I haven't been adding a teardown step.

Comment (ccarouge) [reply tests/test_repository.py:114]: Could we have a comment at the file level to remind us of that? Happy to have the same comment in all `test_*` files. But I am sure to have forgotten in a few days otherwise.

File: tests/test_repository.py (hunk 6 above)
Comment (SeanBryan51) [reply tests/test_repository.py:344]: I'm not sure if `repo` should be an `autouse` fixture since the tests need the `CableRepository` instance that it returns for testing, i.e. we would have to pass it in as an argument to each test regardless.

But yep `remove_module_lines` is probably in the wrong module but given it is only used in this file I haven't bothered to move it yet

File: tests/test_repository.py (hunk 2 above)
Comment (SeanBryan51) [reply tests/test_repository.py:114]: Sure, we can put a comment the module docstring, e.g:
```python
"""`pytest` tests for `repository.py`.

Note: explicit teardown for generated files and directories are not required as
the working directory used for testing is cleaned up in the `_run_around_tests`
pytest autouse fixture.
"""
```

**Review by ccarouge — APPROVED** [review]
If you can add the module docstring before merging that would be great.
